test/bin/build_images.sh

```diff
@@ -270,47 +271,43 @@ do_group() {
         fi
         echo "Building edge-commit from ${blueprint} ${parent_args}"
         # shellcheck disable=SC2086  # quote to avoid glob expansion
-        buildid=$(sudo composer-cli compose start-ostree \
-                       ${parent_args} \
-                       --ref "${blueprint}" \
-                       "${blueprint}" \
-                       edge-commit \
-                      | awk '{print $2}')
+        build_cmd="sudo composer-cli compose start-ostree ${parent_args} --ref ${blueprint} ${blueprint} edge-commit"
+        buildid=$(${build_cmd} | awk '{print $2}')
         echo "Build ID ${buildid}"
         # Record a "build name" to be used as part of the unique
         # filename for the log we download next.
         echo "${blueprint}-edge-commit" > "${IMAGEDIR}/builds/${buildid}.build"
-        buildid_list="${buildid_list} ${buildid}"
+        buildid_list+=("${buildid},${build_cmd}")
     done
 
     if ${BUILD_INSTALLER} && ! ${COMPOSER_DRY_RUN}; then
         for image_installer in "${groupdir}"/*.image-installer; do
             blueprint=$("${GOMPLATE}" --file "${image_installer}")
             echo "Building image-installer from ${blueprint}"
-            buildid=$(sudo composer-cli compose start \
-                           "${blueprint}" \
-                           image-installer \
-                          | awk '{print $2}')
+            build_cmd="sudo composer-cli compose start ${blueprint} image-installer"
+            buildid=$(${build_cmd} | awk '{print $2}')
             echo "Build ID ${buildid}"
             # Record a "build name" to be used as part of the unique
             # filename for the log we download next.
             echo "${blueprint}-image-installer" > "${IMAGEDIR}/builds/${buildid}.build"
-            buildid_list="${buildid_list} ${buildid}"
+            buildid_list+=("${buildid},${build_cmd}")
+            #buildid_list="${buildid_list}#\"${buildid},${build_cmd}\""
         done
     fi
 
-    if [ -n "${buildid_list}" ]; then
+    if (( ${#buildid_list[@]} )); then
         echo "Waiting for builds to complete..."
-        # shellcheck disable=SC2086  # pass command arguments quotes to allow word splitting
-        time "${SCRIPTDIR}/wait_images.py" ${buildid_list}
+        # wait_images.py returns possibly updated list of builds that must be handled
+        # "update" means replacing initial build ID with retry build ID
+        builds_to_get=$(time "${SCRIPTDIR}/wait_images.py" "${buildid_list[@]}")
```

Comment:
Should we better save those builds in a file? Transferring them by stdout seems a bit dangerous if a script fails, etc.

Reply:
File name can be an argument to the wait_images.py script.

Reply:
Do you think the risk is significant that we need to complicate the data flow?

To test I added `sys.exit("test")` before `main(input)`:

```
+ echo 'Waiting for builds to complete...'
Waiting for builds to complete...
++ /home/pm/microshift/test/bin/wait_images.py '487bd1da-33d3-4b0f-9664-8e9ce00afbfb,sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-crel rhel-9.2-microshift-crel edge-commit' 'b8a17c36-5b32-4c50-9cb2-8f56bf8af9f8,sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-base rhel-9.2-microshift-source-base edge-commit' 'd1e32594-cc99-4250-8c59-bc70c527b304,sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-fake-next-minor rhel-9.2-microshift-source-fake-next-minor edge-commit' 'b398a5de-f0c6-42a4-a261-8d7da2185293,sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-fake-yplus2-minor rhel-9.2-microshift-source-fake-yplus2-minor edge-commit' '8bd6b8d5-05e6-4655-8dca-05eefe57d527,sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source rhel-9.2-microshift-source edge-commit'
2023-11-24 10:44:15,209: Received arguments: {'487bd1da-33d3-4b0f-9664-8e9ce00afbfb': 'sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-crel rhel-9.2-microshift-crel edge-commit', 'b8a17c36-5b32-4c50-9cb2-8f56bf8af9f8': 'sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-base rhel-9.2-microshift-source-base edge-commit', 'd1e32594-cc99-4250-8c59-bc70c527b304': 'sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-fake-next-minor rhel-9.2-microshift-source-fake-next-minor edge-commit', 'b398a5de-f0c6-42a4-a261-8d7da2185293': 'sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source-fake-yplus2-minor rhel-9.2-microshift-source-fake-yplus2-minor edge-commit', '8bd6b8d5-05e6-4655-8dca-05eefe57d527': 'sudo composer-cli compose start-ostree --parent rhel-9.2 --url http://192.168.0.10:8080/repo --ref rhel-9.2-microshift-source rhel-9.2-microshift-source edge-commit'}
test

real    0m0.031s
user    0m0.027s
sys     0m0.004s
+ builds_to_get=
+ osbuild_logs
++ sudo systemctl list-units
++ awk '/osbuild-worker@/ {print $1} /osbuild-composer\.service/ {print $1}'
+ workers_services='osbuild-composer.service
osbuild-worker@1.service
osbuild-worker@2.service
osbuild-worker@3.service
osbuild-worker@4.service
osbuild-worker@5.service'
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-composer.service
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-worker@1.service
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-worker@2.service
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-worker@3.service
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-worker@4.service
+ for service in ${workers_services}
+ sudo journalctl -u osbuild-worker@5.service

$ echo $?
1
```

Reply:
OK, so if the script exits with error, the procedure fails - that's good.
I'm not sure if we absolutely need to implement this. I'm just suspicious of program output usage for follow-up actions in general.
Let's decide about it on Monday

Reply:
But we do it all the time, right? Even for starting the jobs we take the UUID from the output.